src/WebJobs.Script.WebHost/Security/KeyManagement/KeyVaultSecretsRepository.cs

```diff
@@ -0,0 +1,298 @@
+                {
+                    hostSecrets.FunctionKeys.Add(SecretBundleToKey(item, functionKeyPrefix));
+                }
+                else if (item.SecretIdentifier.Name.StartsWith(systemKeyPrefix))
+                {
+                    hostSecrets.SystemKeys.Add(SecretBundleToKey(item, systemKeyPrefix));
+                }
+            }
+
+            return hostSecrets;
+        }
+
+        private async Task<ScriptSecrets> ReadFunctionSecrets(string functionName)
+        {
+            IPage<SecretItem> secretItems = await _keyVaultClient.Value.GetSecretsAsync(GetVaultBaseUrl());
+            List<Task<SecretBundle>> tasks = new List<Task<SecretBundle>>();
+            string prefix = $"{functionPrefix}{Normalize(functionName)}--";
+            foreach (SecretItem item in secretItems.Where(x => x.Identifier.Name.StartsWith(prefix, StringComparison.OrdinalIgnoreCase)))
+            {
+                tasks.Add(_keyVaultClient.Value.GetSecretAsync(GetVaultBaseUrl(), item.Identifier.Name));
+            }
+            if (!tasks.Any())
+            {
+                return null;
+            }
+            await Task.WhenAll(tasks);
+
+            FunctionSecrets functionSecrets = new FunctionSecrets()
+            {
+                Keys = new List<Key>()
+            };
+
+            foreach (Task<SecretBundle> task in tasks)
+            {
+                SecretBundle item = task.Result;
+                functionSecrets.Keys.Add(SecretBundleToKey(item, prefix));
+            }
+
+            return functionSecrets;
+        }
+
+        private string GetVaultBaseUrl()
+        {
+            return $"https://{_vaultName}.vault.azure.net";
+        }
+
+        public static Dictionary<string, string> GetDictionaryFromScriptSecrets(ScriptSecrets secrets, string functionName)
+        {
+            Dictionary<string, string> dic = new Dictionary<string, string>();
+            if (secrets is HostSecrets)
+            {
+                HostSecrets hostSecrets = secrets as HostSecrets;
+                if (hostSecrets.MasterKey != null)
+                {
+                    dic.Add(masterKey + "master", hostSecrets.MasterKey.Value);
+                }
+
+                if (hostSecrets.FunctionKeys != null)
+                {
+                    foreach (Key key in hostSecrets.FunctionKeys)
+                    {
+                        dic.Add($"{functionKeyPrefix}{Normalize(key.Name)}", key.Value);
+                    }
+                }
+
+                if (hostSecrets.SystemKeys != null)
+                {
+                    foreach (Key key in hostSecrets.SystemKeys)
+                    {
+                        dic.Add($"{systemKeyPrefix}{Normalize(key.Name)}", key.Value);
+                    }
+                }
+            }
+            else
+            {
+                FunctionSecrets functionSecrets = secrets as FunctionSecrets;
```

Comment:
This implementation assumes `ScriptSecrets` will always be either `HostSecrets` or `FunctionSecrets`, checking that this cast succeeded here would help avoid issues if that changes.